File: src/test/kotlin/school/charset/app/domain/exercise/ValidationResultTest.kt
package school.charset.app.domain.exercise

import io.kotest.core.spec.style.FreeSpec
import io.kotest.matchers.shouldBe

class ValidationResultTest :
    FreeSpec({
        "correct()" - {
            "produces ok=true, no errorType, no params" {
                val result = ValidationResult.correct()
                result.ok shouldBe true
                result.errorType shouldBe null
                result.params shouldBe emptyMap()
            }
        }

        "incorrect()" - {
            "with errorType only produces ok=false and empty params" {
                val result = ValidationResult.incorrect(errorType = "some.error")
                result.ok shouldBe false
                result.errorType shouldBe "some.error"
                result.params shouldBe emptyMap()
            }

            "with errorType and params" {
                val result = ValidationResult.incorrect(
                    errorType = "some.error",
                    params = mapOf("got" to "x"),
                )
                result.ok shouldBe false
                result.errorType shouldBe "some.error"
                result.params shouldBe mapOf("got" to "x")
            }
        }

        "invariants" - {
            // The primary constructor is private, but reflection or a sloppy refactor
            // could still produce invalid states. These tests pin the runtime guarantee.

            "incorrect with empty errorType is impossible via factory" {
                // incorrect() requires a non-null errorType; missing it is a compile error.
                // Smoke test only: factory accepts a string.
                ValidationResult.incorrect("anything").ok shouldBe false
            }
        }

        // Equality / hashCode behavior - needed because validator tests compare via shouldBe.
        "equality" - {
            "two correct() results are equal" {
                ValidationResult.correct() shouldBe ValidationResult.correct()
            }

            "two incorrect() with same errorType and params are equal" {
                ValidationResult.incorrect("a", mapOf("k" to "v")) shouldBe
                    ValidationResult.incorrect("a", mapOf("k" to "v"))
            }

            "correct() != incorrect()" {
                (ValidationResult.correct() == ValidationResult.incorrect("a")) shouldBe false
            }
        }
    })

File: test/charset/exercise/validation_result_test.exs
defmodule Charset.Exercise.ValidationResultTest do
  use ExUnit.Case, async: true

  alias Charset.Exercise.ValidationResult

  describe "correct/0" do
    test "produces ok=true, no error_type, no params" do
      result = ValidationResult.correct()
      assert result.ok
      assert result.error_type == nil
      assert result.params == %{}
    end
  end

  describe "incorrect/2" do
    test "with error_type only produces ok=false and empty params" do
      result = ValidationResult.incorrect("some.error")
      refute result.ok
      assert result.error_type == "some.error"
      assert result.params == %{}
    end

    test "with error_type and params" do
      result = ValidationResult.incorrect("some.error", %{"got" => "x"})
      refute result.ok
      assert result.error_type == "some.error"
      assert result.params == %{"got" => "x"}
    end
  end

  # Equality behavior - the validator tests compare whole structs.
  describe "equality" do
    test "two correct() results are equal" do
      assert ValidationResult.correct() == ValidationResult.correct()
    end

    test "two incorrect() with same error_type and params are equal" do
      assert ValidationResult.incorrect("a", %{"k" => "v"}) ==
               ValidationResult.incorrect("a", %{"k" => "v"})
    end

    test "correct() != incorrect()" do
      refute ValidationResult.correct() == ValidationResult.incorrect("a")
    end
  end
end
